internal/engine/upper.go

```diff
@@ -92,9 +92,21 @@ func NewUpper(ctx context.Context, b BuildAndDeployer, k8s k8s.Client,
 	}
 }
 
+func (u Upper) NewLogActionLogger(ctx context.Context) logger.Logger {
+	l := logger.Get(ctx)
+	return logger.NewFuncLogger(l.SupportsColor(), l.Level(), func(level logger.Level, b []byte) error {
+		u.store.Dispatch(LogAction{b})
+		return nil
+	})
+}
+
 func (u Upper) RunHud(ctx context.Context) error {
+	state := u.store.LockMutableState()
```

Comment:
this seems...weird? can we have Hud dispatch an action when it starts running and move this into the reducers?